snippets/base/management/commands/fetch_daily_metrics.py

```diff
@@ -0,0 +1,109 @@
+from datetime import datetime, timedelta
+
+from django.conf import settings
+from django.core.management.base import BaseCommand, CommandError
+from django.db.models import Q
+
+import sentry_sdk
+from redash_dynamic_query import RedashDynamicQuery
+
+from snippets.base.models import DailyJobMetrics, Job
+
+
+class Command(BaseCommand):
+    args = "(no args)"
+    help = "Fetch daily Job metrics"
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--date',
+            help='Fetch data for date. Defaults to yesterday. In YYYYMMDD format.',
+        )
+
+    def handle(self, *args, **options):
+        if not options['date']:
+            now = datetime.utcnow()
+            date = (
+                datetime(year=now.year, month=now.month, day=now.day, hour=0, minute=0, second=0) -
+                timedelta(days=1)
+            )
+        else:
+            date = datetime.strptime(options['date'], '%Y%m%d')
+
+        last_week = date - timedelta(days=7)  # date - 7 since we 're fetching data for yesterday.
+        date_next_day = date + timedelta(days=1)
+
+        if not settings.REDASH_API_KEY:
+            raise CommandError('Enviroment variable REDASH_API_KEY is required.')
+
+        redash = RedashDynamicQuery(
+            endpoint=settings.REDASH_ENDPOINT,
+            apikey=settings.REDASH_API_KEY,
+            max_wait=settings.REDASH_MAX_WAIT,
+        )
+        jobs = (
+            Job.objects
+            .filter(status__in=[Job.PUBLISHED, Job.COMPLETED])
+            .filter(
+                # Publish start before the end of day of interest and still
+                # not completed.
+                (Q(publish_start__lt=date_next_day) & Q(completed_on=None)) |
+                # Or completed during the last 7 days before EOD of the day of interest.
+                (Q(completed_on__gte=last_week) & Q(completed_on__lt=date_next_day))
+            )
+        )
+
+        self.stdout.write(f'Fetching data for {date} for {jobs.count()} jobs.')
+
+        data_fetched = False
+        for job in jobs:
+            bind_data = {
+                'start_date': date.strftime('%Y%m%d'),
+                'end_date': date.strftime('%Y%m%d'),
+                'message_id': job.id,
+            }
+            try:
+                result = redash.query(settings.REDASH_QUERY_ID, bind_data)
```

Comment:
I'm wondering about the performance implications of doing a separate redash query for each job. For the small number of jobs published each day this probably doesn't matter, but for loading all of the data going back to when we started publish jobs we might want to use a different approach. We can follow up with a separate script for that if needed though.

Reply:
We're going to load historical data starting from Q4, since we have incomplete data for Sept and Q3 anyway. I constructed the command in this way so we can import which day separately. With a loop we will fetch all data within 24 hours which is ok for a one-off.

> For the small number of jobs published each day

Each day we need to fetch data for more than 1000 Jobs and we need about 90 minutes to do so.

Reply:
Actually what @jgmize proposed was way smarter and implemented things that way. Queries only take a few seconds per day now.